<?xml version="1.0"?>

<launch>

  <env name="ROSCONSOLE_CONFIG_FILE" value="$(find robotx_gazebo)/config/custom_rosconsole.conf"/>

  <!-- For setting debug level to debug -->
  <!--<env name="ROSCONSOLE_CONFIG_FILE"
       value="$(find kingfisher_gazebo)/config/custom_rosconsole.conf"/>
   -->
  <!-- Command line arguments -->

  <arg name="world" default="bamboolake_bionic.world"/>

  <arg name="name" default="WAMV"/>

  <!-- Initial USV location and attitude-->
  <arg name="x" default="-5" />
  <arg name="y" default="5" />
  <arg name="z" default="0.1" />
  <arg name="P" default="0" />
  <arg name="R" default="0" />
  <arg name="Y" default="0" />
  <arg name="down_edge" default="0"/>
  <arg name="left_edge" default="0"/>
  <arg name="top_edge" default="20"/>
  <arg name="right_edge" default="20"/>


  <!-- The switch about sensors and wind -->
  <arg name="front_camera" default="true"/>
  <arg name="gps" default="true"/>
  <arg name="imu" default="true"/>
  <arg name="lidar" default="true"/>
  <arg name="wind" default="true"/> 
  <arg name="p3d" default="false"/> 

  <!-- The switch about apriltags on WAM-V -->
  <arg name="apriltags" default="true"/> 

  <param name="use_sim_time" value="true" />
  <!-- Start Gazebo with the world file -->
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="$(find robotx_gazebo)/worlds/$(arg world)"/>
    <arg name="verbose" value="true"/>
    <arg name="paused" value="true"/>
    <arg name="use_sim_time" value="true"/>
    <arg name="gui" value="true"/>
    <arg name="headless" value="false"/>
  </include>

  <!-- Load WAM-V robot model -->
  <arg name="urdf" default="$(find wamv_description)/urdf/wamv_robotx_nctu.xacro"/>
  <!--<arg name="urdf" default="$(find nps_robotx)/urdf/nps_wamv.xacro"/> -->
  <param name="robot_description" command="$(find xacro)/xacro --inorder $(arg urdf) wind:=$(arg wind) 
  front_camera:=false lidar:=false gps:=$(arg gps) imu:=$(arg imu)
   p3d:=true apriltags:=$(arg apriltags)"/>
  <!-- Spawn model in Gazebo -->
  <node name="spawn_model" pkg="gazebo_ros" type="spawn_model"
  args="-x $(arg x) -y $(arg y) -z $(arg z)
        -R $(arg R) -P $(arg P) -Y $(arg Y)
        -unpause -urdf -param robot_description -model wamv"/>

  <!-- Load watchtower robot model -->
  <!-- down -->
  <arg name="urdf_watchtower_down" default="$(find wamv_description)/urdf/watchtower.xacro"/>
  <param name="watchtower_down_description" command="$(find xacro)/xacro --inorder $(arg urdf_watchtower_down) wind:=$(arg wind) lidar:=false p3d:=$(arg p3d) robot_name:=watchtower_down"/>
  <node name="spawn_model_watchtower_down" pkg="gazebo_ros" type="spawn_model"
  args="-x 10 -y -12 -z $(arg z)
        -R $(arg R) -P $(arg P) -Y 1.57079
        -unpause -urdf -param watchtower_down_description -model watchtower_down"/>

  <!-- left -->
  <arg name="urdf_watchtower_left" default="$(find wamv_description)/urdf/watchtower.xacro"/>
  <param name="watchtower_left_description" command="$(find xacro)/xacro --inorder $(arg urdf_watchtower_left) wind:=$(arg wind) lidar:=false p3d:=$(arg p3d) robot_name:=watchtower_left"/>
  <node name="spawn_model_watchtower_left" pkg="gazebo_ros" type="spawn_model"
  args="-x -12 -y 10 -z $(arg z)
        -R $(arg R) -P $(arg P) -Y 0
        -unpause -urdf -param watchtower_left_description -model watchtower_left"/>

  <!-- Load apriltag robot model  -->
    
  <!-- down_left -->
  <arg name="urdf_apriltag_down_left" default="$(find wamv_description)/urdf/apriltag.xacro"/>
  <param name="apriltag_down_left_description" command="$(find xacro)/xacro --inorder $(arg urdf_apriltag_down_left) wind:=$(arg wind) p3d:=$(arg p3d) robot_name:=apriltag_down_left tag_id:=504 tag_id2:=508 tag_size:=1.0"/>
  <node name="spawn_model_apriltag_down_left" pkg="gazebo_ros" type="spawn_model"
  args="-x $(arg left_edge) -y $(arg down_edge) -z $(arg z)
        -R $(arg R) -P $(arg P) -Y 1.57079
        -unpause -urdf -param apriltag_down_left_description -model apriltag_down_left"/>

  <!-- top_left -->
  <arg name="urdf_apriltag_top_left" default="$(find wamv_description)/urdf/apriltag.xacro"/>
  <param name="apriltag_top_left_description" command="$(find xacro)/xacro --inorder $(arg urdf_apriltag_top_left) wind:=$(arg wind) p3d:=$(arg p3d) robot_name:=apriltag_top_left tag_id:=506 tag_id2:=510 tag_size:=1.0"/>
  <node name="spawn_model_apriltag_top_left" pkg="gazebo_ros" type="spawn_model"
  args="-x $(arg left_edge) -y $(arg top_edge) -z $(arg z)
        -R $(arg R) -P $(arg P) -Y 1.57079
        -unpause -urdf -param apriltag_top_left_description -model apriltag_top_left"/>

  <!-- top_right -->
  <arg name="urdf_apriltag_top_right" default="$(find wamv_description)/urdf/apriltag.xacro"/>
  <param name="apriltag_top_right_description" command="$(find xacro)/xacro --inorder $(arg urdf_apriltag_top_right) wind:=$(arg wind) p3d:=$(arg p3d) robot_name:=apriltag_top_right tag_id:=507 tag_id2:=511 tag_size:=1.0"/>
  <node name="spawn_model_apriltag_top_right" pkg="gazebo_ros" type="spawn_model"
  args="-x $(arg right_edge) -y $(arg top_edge) -z $(arg z)
        -R $(arg R) -P $(arg P) -Y 1.57079
        -unpause -urdf -param apriltag_top_right_description -model apriltag_top_right"/>

  <!-- down_right -->
  <arg name="urdf_apriltag_down_right" default="$(find wamv_description)/urdf/apriltag.xacro"/>
  <param name="apriltag_down_right_description" command="$(find xacro)/xacro --inorder $(arg urdf_apriltag_down_right) wind:=$(arg wind) p3d:=$(arg p3d) robot_name:=apriltag_down_right tag_id:=505 tag_id2:=509 tag_size:=1.0"/>
  <node name="spawn_model_apriltag_down_right" pkg="gazebo_ros" type="spawn_model"
  args="-x $(arg right_edge) -y $(arg down_edge) -z $(arg z)
        -R $(arg R) -P $(arg P) -Y 1.5707
        -unpause -urdf -param apriltag_down_right_description -model apriltag_down_right"/>



  <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher">
    <param name="publish_frequency" type="double" value="10.0" />
  </node>

  <!-- run waypoint navigation contoller -->
  <node pkg="robotx_gazebo" type="WAMV_PID_controller.py" name="wamv_pid_controller">
  </node>

  <node pkg="robotx_gazebo" type="wamv_model_path.py" name="wamv_model_path"/>

  <!--<include file="$(find robotx_gazebo)/launch/usv_diffdrive.launch"/>-->

  <!-- Utility for publishing rpy for plotting -->
  <node pkg="robotx_gazebo" type="quat2rpy.py" name="quat2rpy" output="screen">
      <remap from="in_topic" to="/gazebo/model_states" />
      <remap from="out_topic" to="wamv_rpy" />
      <param name="input_msg_type" value="ModelStates" />
      <param name="modelstates_index" value="1" type="int" />
  </node>

</launch>
